src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java

```diff
@@ -94,17 +102,50 @@ private static void ensureNoInputConflicts(CartesianProductTest annotation,
 		return sets;
 	}
 
-	private List<List<?>> getSetsFromRepeatableAnnotation(List<CartesianValueSource> valueSources) {
+	private List<List<?>> getSetsFromArgumentsSources(List<? extends Annotation> argumentsSources,
+			ExtensionContext context) {
 		List<List<?>> sets = new ArrayList<>();
-		for (CartesianValueSource source : valueSources) {
-			CartesianValueArgumentsProvider provider = new CartesianValueArgumentsProvider();
-			provider.accept(source);
-			List<Object> collect = provider.provideArguments().distinct().collect(toList());
-			sets.add(collect);
+		for (Annotation source : argumentsSources) {
+			sets.add(getSetFromAnnotation(context, source));
 		}
 		return sets;
 	}
 
+	private List<Object> getSetFromAnnotation(ExtensionContext context, Annotation source) {
+		try {
+			ArgumentsProvider provider = initializeArgumentsProvider(source);
+			return provideArguments(context, source, provider);
+		}
+		catch (Exception e) {
+			throw new ExtensionConfigurationException("Could not provide arguments because of exception.", e);
+		}
+	}
+
+	private ArgumentsProvider initializeArgumentsProvider(Annotation source) {
+		ArgumentsSource providerAnnotation = AnnotationSupport
+				.findAnnotation(source.annotationType(), ArgumentsSource.class)
+				// never happens, we already know these annotations are annotated with @ArgumentsSource
+				.orElseThrow(() -> new PreconditionViolationException(format(
+					"%s was not annotated with @ArgumentsSource but should have been.", source.annotationType())));
+		return ReflectionSupport.newInstance(providerAnnotation.value());
+	}
+
+	private List<Object> provideArguments(ExtensionContext context, Annotation source, ArgumentsProvider provider)
+			throws Exception {
+		if (Arrays.asList(provider.getClass().getInterfaces()).contains(Consumer.class)) {
```

Comment:
This should be...

```suggestion
		if (provider instanceof Consumer) {
```

... but, it's not? (One test fails after the change.)

Reply:
`ArgumentsProvider`s used by annotations belonging to JUnit's `@ParameterizedTest` (e.g.: `@ValueSource`) implement `Consumer` through `AnnotationConsumer`. We will fix this with defining `CartesianAnnotationConsumer`.